lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -695,6 +701,39 @@ deleteCheckpoint wid sid = do
     deleteWhere [CheckpointWalletId ==. wid, CheckpointSlot ==. sid]
     deleteCheckpointState @s wid sid
 
+-- | Clean up checkpoints which are greater than `k` blocks old for we know we
+-- can't rollback that far.
+purgeCheckpoints
+    :: W.WalletId
+    -> W.Wallet s t
+    -> SqlPersistT IO ()
+purgeCheckpoints wid cp = do
+    let minHeight = word64 (blockHeight cp) - epochStability
+    mCp <- selectFirst
+        [ CheckpointWalletId ==. wid
+        , CheckpointBlockHeight <=. minHeight
+        ] [Desc CheckpointBlockHeight]
+    case mCp of
+        Nothing -> return ()
+        Just minCp -> do
+            deleteWhere
+                [ UtxoWalletId ==. wid
+                , UtxoSlotSpent !=. Nothing
+                , UtxoSlotSpent <=. Just (checkpointSlot $ entityVal minCp)
```

Comment:
we have the following situation here
`slot     row in db`
0. slot0
1. slot1 -> Utxo(slot1,Nothing)
2. slot2 -> Utxo(slot1,Nothing)
3. slot3 -> Utxo(slot1,Just(slot3))
4. slot4 -> Utxo(slot1,Just(slot3))
Here we keep at slot4 this object as it can be needed in rollback
Let's say we start to purge
a. we purge checkpoint put at slot 0 -> Utxo(slot1,Just(slot3)
b. we purge checkpoint put at slot 1 -> Utxo(slot1,Just(slot3)
c. we purge checkpoint put at slot 2 -> Utxo(slot1,Just(slot3)
d. we purge checkpoint put at slot 3 -> no object anymore
e. we purge checkpoint put at slot 4 -> no object anymore
we will not do rollback deeper than purge - this is how I understand it

Reply:
side note: the logic will get more complicated if k changes (especially if it increases and then very deep rollback happens). in such a case we will need to restore some blocks to support it

Reply:
Not much actually. If k suddenly shrinks, well good, we'll do another purge.
if k increases, then it's likely that we have purged too much and we'll steadily catch-up until we start having enough checkpoints again. In the event of a deep rollback (unlikely), we'll have to rollback from Genesis. 

This is "unfortunate" but, that's something which will most likely never happen or, in a very extreme case. If so, then the wallet will be restoring for a little longer than usual :)

Reply:
Also, purging isn't necessary so to speak. And, we could be more lenient about it and only purge things that are beyond 10*k blocks for instances.

Reply:
indeed, we can have `security paameter *2`

Reply:
What we do in future about _k_ increasing also depends on how the node works in that circumstance (in the future).

You are right that purging isn't necessary... We may be able to add a cheap optimisation to only do a purge if `slot mod 97 = 0`.

Reply:
`97` as an arbitrary prime number which is big enough?

> depends on how the node works in that circumstance

What do you in mind here :thinking:?